### PR TITLE
Enable caching on intermediate realtime persists

### PR DESCRIPTION
**Pull request by xvrl**

### CODE REVIEW

File: server/src/main/java/io/druid/client/cache/MemcachedCache.java [hunk 1]
```diff
@@ -26,6 +26,7 @@
 import com.google.common.base.Throwables;
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
```
Comment (nishantmonu51) [inline server/src/main/java/io/druid/client/cache/MemcachedCache.java]: unused import ?

File: server/src/main/java/io/druid/segment/realtime/plumber/RealtimePlumber.java [hunk 2]
```diff
@@ -901,6 +930,16 @@ protected void abandonSegment(final long truncatedTime, final Sink sink)
...
+      Iterables.transform(sink, new Function<FireHydrant, Void>()
+      {
+        @Nullable
+        @Override
+        public Void apply(@Nullable FireHydrant input)
+        {
+          cache.close(input.getSegment().getIdentifier() + "_" + input.getCount());
+          return null;
+        }
+      });
```
Comment (nishantmonu51) [inline server/src/main/java/io/druid/segment/realtime/plumber/RealtimePlumber.java]: do we need to iterate for the close to be called ? or may be simply replace it with a for loop. 
Also, can we abstract the the namespace creation method.

File: server/src/main/java/io/druid/segment/realtime/plumber/RealtimePlumber.java [hunk 3]
```diff
@@ -901,6 +932,16 @@ protected void abandonSegment(final long truncatedTime, final Sink sink)
...
+      Iterables.transform(sink, new Function<FireHydrant, Void>()
+      {
+        @Nullable
+        @Override
+        public Void apply(@Nullable FireHydrant input)
+        {
+          cache.close(makeHydrantIdentifier(input, input.getSegment()));
+          return null;
+        }
+      });
```
Comment (nishantmonu51) [inline server/src/main/java/io/druid/segment/realtime/plumber/RealtimePlumber.java]: do we need to iterate for the close to be called ? or may be simply replace it with a for loop. 
Also, can we abstract the the namespace creation method.

Comment (gianm) [inline server/src/main/java/io/druid/segment/realtime/plumber/RealtimePlumber.java]: `Iterables.transform` is lazy so I think this code does not actually do anything.

File: server/src/main/java/io/druid/segment/realtime/plumber/RealtimePlumber.java [hunk 4]
```diff
@@ -353,6 +382,8 @@ private Sink getSink(long timestamp)
     );
   }
 
+  protected static String makeHydrantIdentifier(FireHydrant input, ReferenceCountingSegment segment) {return segment.getIdentifier() + "_" + input.getCount();}
```
Comment (gianm) [inline server/src/main/java/io/druid/segment/realtime/plumber/RealtimePlumber.java]: Formatting

File: server/src/main/java/io/druid/segment/realtime/plumber/RealtimePlumber.java (hunk 3 above)
Comment (xvrl) [inline server/src/main/java/io/druid/segment/realtime/plumber/RealtimePlumber.java]: yep, fixing that already

**Comment by gianm** [comment]
The spills are not consistent across nodes so this would only work if the cache is local. That should probably be called out or enforced or something.

**Comment by gianm** [comment]
@xvrl is it possible to get real world performance numbers?

**Comment by xvrl** [comment]
@gianm addressed comments, will look into getting some real-work numbers.

**Comment by xvrl** [comment]
```

**Comment by xvrl** [comment]
```

**Comment by xvrl** [comment]
@gianm roughly 20% speedup on a typical top-n query 10-15min into the hour of a 1-hour realtime index task, with 10 persists completed (~ 1.4M rows to scan)

```
Unit: milliseconds
           expr      min       lq     mean   median       uq      max neval
  cache("true") 226.2689 269.5347 290.6316 279.4486 302.8039 542.1334   100
 cache("false") 285.9610 343.3580 395.8290 396.8076 419.2823 708.2887   100
```

**Comment by himanshug** [comment]
:+1: since old behavior is preserved by default and this gets into action only if user explicitly provides the configuration.
not sure if documentation update is left out intentionally , if not then pls update same.

**Comment by xvrl** [comment]
@himanshug yes, will add docs, was waiting for comments before doing so

**Pull request merged**